src/object_store/src/object/s3.rs

```diff
@@ -12,17 +12,99 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use aws_sdk_s3::model::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::output::UploadPartOutput;
 use aws_sdk_s3::{Client, Endpoint, Region};
 use fail::fail_point;
 use futures::future::try_join_all;
 use itertools::Itertools;
+use tokio::sync::Mutex;
 
-use super::{BlockLocation, ObjectError, ObjectMetadata};
+use super::{BlockLocation, MultipartUploadHandle, ObjectError, ObjectMetadata, PartId};
 use crate::object::{Bytes, ObjectResult, ObjectStore};
 
+/// S3 multipart upload handle.
+/// Reference: <https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html>
+struct UploadHandle {
+    client: Arc<Client>,
+    bucket: String,
+    /// The key of the object.
+    key: String,
+    /// The identifier of multipart upload task for S3.
+    upload_id: String,
+    /// Parts that are already uploaded to S3.
+    /// The bool member indicates whether the upload is finished.
+    uploaded_parts: Mutex<(bool, BTreeMap<i32, UploadPartOutput>)>,
+}
+
+#[async_trait::async_trait]
+impl MultipartUploadHandle for UploadHandle {
+    async fn upload_part(&self, part_id: PartId, part: Bytes) -> ObjectResult<()> {
+        fail_point!("s3_upload_part_err", |_| Err(ObjectError::internal(
```

Comment:
maybe just: s3_upload_part_err -> s3_upload_part

Reply:
The rest of the fail points all used `s3_xxx_err`

Reply:
Perhaps I should change `s3_finish_multipart_upload` to `s3_finish_multipart_upload_err`.